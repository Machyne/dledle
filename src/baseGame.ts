// Give these string values so they can be round trip through being used as object keys.
// These are never encoded as-is so strings are fine.
export enum GameScore {
  Win = "Win",
  NearWin = "NearWin",
  Loss = "Loss",
}

export type GameResult = {
  score: GameScore;
  serializedResult: string;
};

export abstract class BaseGame {
  private _resultRegex: RegExp | null = null;

  /**
   * The name of the game.
   */
  abstract get name(): string;

  /**
   * The name of the game in lowercase with only alphanumeric characters.
   */
  get fileName(): string {
    return this.name.toLowerCase().replace(/[^a-z0-9]/g, "");
  }

  /**
   * A regex that matches the result of the game.
   */
  get resultRegex(): RegExp {
    if (this._resultRegex === null) {
      this._resultRegex = this._buildResultRegex();
    }
    return this._resultRegex;
  }

  /**
   * Builds a regex that matches the result of the game. The return value will be cached.
   */
  abstract _buildResultRegex(): RegExp;

  /**
   * A link to the game.
   */
  abstract get link(): string;

  /**
   * Parses the user-supplied game result and returns a serialized version of
   * the game data in addition to the "score".
   *
   * @param gameResult the result of the game run through the resultRegex
   */
  abstract serializeResult(gameResult: RegExpMatchArray): GameResult;

  /**
   * The share text of the result of the game given the serialized game data.
   *
   * @param serializedResult the serialized game data
   */
  abstract deserialize(serializedResult: string): string;
}
