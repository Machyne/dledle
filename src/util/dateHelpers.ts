export type DateParts = {
  year: number;
  month: number;
  day: number;
};

export function ymd(year: number, month: number, day: number): DateParts {
  return { year, month, day };
}
export const FIRST_GAME_DATE: DateParts = ymd(2023, 12, 21);

/*
 * Returns the number of days between two dates (ie `a - b` in days).
 * If a is earlier than b, the result will be negative.
 */
export function signedDayDifference(a: DateParts, b: DateParts): number {
  const aDate = new Date(a.year, a.month - 1, a.day);
  const bDate = new Date(b.year, b.month - 1, b.day);
  return Math.round((aDate.getTime() - bDate.getTime()) / (1000 * 60 * 60 * 24));
}

export function gameNumberForDate(date: DateParts): number {
  // Add 1 because the first game is game 1, not game 0.
  return signedDayDifference(date, FIRST_GAME_DATE) + 1;
}

export function currentGameNumber(): number {
  return gameNumberForDate(currentOfficialDate());
}

export function currentOfficialDate(): DateParts {
  const formatter = new Intl.DateTimeFormat("en-US", {
    timeZone: "America/Chicago",
    year: "numeric",
    month: "numeric",
    day: "numeric",
  });
  const { year, month, day } = Object.fromEntries(
    formatter.formatToParts(new Date()).map(({ type, value }) => [type, parseInt(value)]),
  );
  return { year, month, day };
}
