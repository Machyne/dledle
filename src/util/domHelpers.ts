export function baseUrl() {
  let port = window.location.port;
  if (port) {
    port = ":" + port;
  }
  return window.location.protocol + "//" + window.location.hostname + port + "/";
}

export function el<T extends HTMLElement = HTMLElement>(id: string): T | null {
  const e = document.getElementById(id) as T | null;
  if (!e) {
    console.error(`No element with id ${id}`);
  }
  return e;
}

export function elOrThrow<T extends HTMLElement = HTMLElement>(id: string): T {
  const e = document.getElementById(id) as T | null;
  if (!e) {
    throw new Error(`No element with id ${id}`);
  }
  return e;
}

export function show(elementOrID: string | Element | null) {
  if (!elementOrID) {
    return;
  }
  if (typeof elementOrID === "string") {
    elementOrID = elOrThrow(elementOrID);
  }
  elementOrID.classList.remove("hidden");
}

export function hide(elementOrID: string | Element | null) {
  if (!elementOrID) {
    return;
  }
  if (typeof elementOrID === "string") {
    elementOrID = elOrThrow(elementOrID);
  }
  elementOrID.classList.add("hidden");
}

export function addHideListener(element: Element, callback: () => void) {
  let wasHidden = element.classList.contains("hidden");
  const attrObserver = new MutationObserver((mutations) => {
    mutations.forEach((mu) => {
      if (mu.type !== "attributes" && mu.attributeName !== "class") return;
      const currentlyHidden = element.classList.contains("hidden");
      if (!wasHidden && currentlyHidden) {
        callback();
      }
      wasHidden = currentlyHidden;
    });
  });
  attrObserver.observe(element, { attributes: true });
}
