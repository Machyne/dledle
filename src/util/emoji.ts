import Graphemer from "graphemer";

const splitter = new Graphemer();

export function splitEmoji(str: string): Array<string> {
  return splitter.splitGraphemes(str).filter((c) => !/^\s+$/u.test(c));
}

export function splitEmojiLines(str: string): Array<Array<string>> {
  const lines = str.split(/\n/u);
  const splitLines = lines.map((line) => splitEmoji(line));
  return splitLines.filter((line) => line.length > 0);
}

export function emojiToRegexUnion(emoji: Array<string>) {
  return `(?:${emoji.join("|")})`;
}

export const cardinalArrowString = "⬆️➡️⬇️⬅️";
export const ordinalArrowString = cardinalArrowString + "↗️↘️↙️↖️";
export const cardinalArrows = splitEmoji(cardinalArrowString);
export const ordinalArrows = splitEmoji(ordinalArrowString);
export const ordinalArrowValues: Record<string, number> = Object.fromEntries(
  ordinalArrows.map((key, idx) => [key, idx]),
);
export const ordinalArrowRegex = emojiToRegexUnion(ordinalArrows);
export const arrows = {
  u: "⬆️" as const,
  r: "➡️" as const,
  d: "⬇️" as const,
  l: "⬅️" as const,
  ur: "↗️" as const,
  dr: "↘️" as const,
  dl: "↙️" as const,
  ul: "↖️" as const,
};

export const emojiSquares = {
  black: "⬛" as const,
  white: "⬜" as const,
  yellow: "🟨" as const,
  green: "🟩" as const,
  red: "🟥" as const,
  purple: "🟪" as const,
};
export const fiveSquaresRegex = "(?:⬛|⬜|🟨|🟩){5}";

export const emojiCircles = {
  white: "⚪" as const,
  yellow: "🟡" as const,
  green: "🟢" as const,
  red: "🔴" as const,
};

export const listToValueMap = (list: Array<string>) =>
  Object.fromEntries(list.map((key, idx) => [key, idx]));

const numbers = "0️⃣1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣8️⃣9️⃣";
export const numberEmoji = splitEmoji(numbers);
export const numberEmojiRegex = emojiToRegexUnion(numberEmoji);
export const numberEmojiValues = listToValueMap(numberEmoji);
export function emojiToNumber(emoji: Array<string>) {
  return emoji.reduce((sum, numEmoji) => 10 * sum + numberEmojiValues[numEmoji], 0);
}

export function numberToEmoji(number: number, padZeros = 0) {
  if (number === 0) {
    return numberEmoji[0].repeat(Math.max(padZeros, 1));
  }
  const result: Array<string> = [];
  while (number > 0) {
    result.push(numberEmoji[number % 10]);
    number = Math.floor(number / 10);
  }
  const numZeros = padZeros - result.length;
  if (numZeros > 0) {
    result.push(numberEmoji[0].repeat(numZeros));
  }
  return result.reverse().join("");
}
