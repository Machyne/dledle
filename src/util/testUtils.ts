import { BaseGame, GameScore } from "../baseGame";

export type TestCases = {
  name: string;
  input: string;
  expectedScore: GameScore;
  expectedOutput?: string;
};

export function runGameTests(game: BaseGame, testCases: Array<TestCases>) {
  const winCase = testCases.find(({ expectedScore }) => expectedScore === GameScore.Win);
  if (winCase) {
    testCases.push({
      name: winCase.name + " with extra link text",
      input: "whatever text\n" + winCase.input + "\n" + game.link,
      expectedScore: GameScore.Win,
      expectedOutput: winCase.expectedOutput ?? winCase.input.trim(),
    });
    testCases.push({
      ...winCase,
      name: winCase.name + " trimmed",
      input: winCase.input.trim(),
    });
  }

  for (const { name, input, expectedScore, expectedOutput } of testCases) {
    it(`should serialize and deserialize ${name} correctly`, () => {
      const match = game.resultRegex.exec(input);
      expect(match).not.toBeNull();
      const { score, serializedResult } = game.serializeResult(match!);

      expect(score).toEqual(expectedScore);
      expect(game.deserialize(serializedResult)).toEqual(expectedOutput ?? input.trim());
    });
  }

  it("serialized data should match snapshot", () => {
    const data = Object.fromEntries(
      testCases.map(({ name, input }) => {
        const match = game.resultRegex.exec(input);
        expect(match).not.toBeNull();
        return [name, game.serializeResult(match!)];
      }),
    );
    expect(data).toMatchSnapshot();
  });
}
