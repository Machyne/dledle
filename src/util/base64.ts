import { DateParts } from "./dateHelpers";

// Feels like a good compromise between density and usability.
// 64^3 = 262,144 which is bigger than 3^10 = 59,049 or 101^2 = 10,201 so it can fit
// two rows of five 3-option guesses or two rows of 0-100 scores.
export const DEFAULT_WORD_WIDTH = 3;

// JS numbers are floats, with 52 bits of mantissa. This means we can represent ints with full precision up to 52 bits long (inclusive).
// This is plenty and it's easier to do it this way; no need to use BigInts.
// Math.floor(JS_MANTISSA_BITS / Math.log2(64)) = ⌊52/6⌋ = 8
const MAX_WORD_WIDTH = 8;

// From https://datatracker.ietf.org/doc/html/rfc4648#section-5
// Table 2: The "URL and Filename safe" Base 64 Alphabet
const B64_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";

const maxNumber = (wordWidth: number) => (1 << (6 * wordWidth)) - 1;

export function intToBase64(n: number, wordWidth = DEFAULT_WORD_WIDTH) {
  if (wordWidth > MAX_WORD_WIDTH) {
    throw new Error(`Word width ${wordWidth} is too large.`);
  }
  if (n < 0) {
    throw new Error(`${n} is negative.`);
  }
  if (n > maxNumber(wordWidth)) {
    throw new Error(`${n} is too large to fit in ${wordWidth} b64 characters.`);
  }

  let result = "";
  for (let i = 0; i < wordWidth; i++) {
    const b64Index = n % 64;
    n = Math.floor(n / 64);
    result = result + B64_ALPHABET[b64Index];
  }
  return result;
}

export function base64ToInt(s: string, wordWidth = DEFAULT_WORD_WIDTH) {
  if (wordWidth > MAX_WORD_WIDTH) {
    throw new Error(`Word width ${wordWidth} is too large.`);
  }
  if (s.length < wordWidth) {
    throw new Error(`${s} is not ${wordWidth} characters long.`);
  }
  let result = 0;
  for (let i = 0; i < wordWidth; i++) {
    var b64Index = B64_ALPHABET.indexOf(s[i]);
    if (b64Index === -1) {
      throw new Error(`${s[i]} is not a valid base64 character.`);
    }
    result += b64Index * 64 ** i;
  }
  return result;
}

export function nextInt(s: string, wordWidth = DEFAULT_WORD_WIDTH): [number, string] {
  if (s.length < wordWidth) {
    throw new Error(`${s} is not ${wordWidth} characters long.`);
  }
  const result = base64ToInt(s, wordWidth);
  return [result, s.slice(wordWidth)];
}

const startYear = 1990;
const maxYear = startYear + Math.floor(maxNumber(DEFAULT_WORD_WIDTH) / (31 * 12)) - 1;

export function datePartsToBase64(d: DateParts) {
  if (d.year < startYear || d.year > maxYear) {
    throw new Error(`Invalid year: ${d.year}`);
  }
  return intToBase64((d.year - startYear) * 31 * 12 + (d.month - 1) * 31 + (d.day - 1));
}

export function base64ToDateParts(s: string): DateParts {
  let intResult = base64ToInt(s);
  const day = (intResult % 31) + 1;
  intResult = Math.floor(intResult / 31);
  const month = (intResult % 12) + 1;
  intResult = Math.floor(intResult / 12);
  const year = intResult + startYear;
  return { year, month, day };
}

export function nextDateParts(s: string): [DateParts, string] {
  if (s.length < DEFAULT_WORD_WIDTH) {
    throw new Error(`${s} is not ${DEFAULT_WORD_WIDTH} characters long.`);
  }
  const result = base64ToDateParts(s);
  return [result, s.slice(DEFAULT_WORD_WIDTH)];
}

const shortStartYear = 2022;
// This gives a max of 2064
const shortMaxYear =
  shortStartYear + Math.floor(maxNumber(DEFAULT_WORD_WIDTH) / 16 / (31 * 12)) - 1;

export function shortDateToBase64(d: DateParts, extra4Bits: number) {
  if (d.year < shortStartYear || d.year > shortMaxYear) {
    throw new Error(`Invalid year: ${d.year}`);
  }
  if (extra4Bits < 0 || extra4Bits > 15) {
    throw new Error(`Invalid extra 4 bits: ${extra4Bits}`);
  }
  const dateNum = (d.year - shortStartYear) * 31 * 12 + (d.month - 1) * 31 + (d.day - 1);
  return intToBase64(dateNum * 16 + extra4Bits, DEFAULT_WORD_WIDTH);
}

export function base64ToShortDate(s: string): [DateParts, number] {
  let intResult = base64ToInt(s);
  const extra4Bits = intResult % 16;
  intResult = Math.floor(intResult / 16);
  const day = (intResult % 31) + 1;
  intResult = Math.floor(intResult / 31);
  const month = (intResult % 12) + 1;
  intResult = Math.floor(intResult / 12);
  const year = intResult + shortStartYear;
  return [{ year, month, day }, extra4Bits];
}

export function nextShortDate(s: string): [DateParts, string, number] {
  if (s.length < DEFAULT_WORD_WIDTH) {
    throw new Error(`${s} is not ${DEFAULT_WORD_WIDTH} characters long.`);
  }
  const result = base64ToShortDate(s);
  return [result[0], s.slice(DEFAULT_WORD_WIDTH), result[1]];
}
