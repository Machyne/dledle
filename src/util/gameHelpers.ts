import { DEFAULT_WORD_WIDTH, intToBase64, nextInt } from "./base64";
import { emojiSquares } from "./emoji";

export function emojiSquareScaleToNumber(guessChars: Array<string>) {
  let result = 0;
  if (guessChars.length < 5) {
    throw new Error(`Invalid guess: ${guessChars}`);
  }
  for (const guessChar of guessChars) {
    if (guessChar === emojiSquares.green) {
      result += 2;
    } else if (guessChar === emojiSquares.yellow) {
      result += 1;
    }
  }
  return result;
}

export function numberToEmojiSquareScale(num: number, padChar: string = emojiSquares.black) {
  if (num < 0 || num > 10) {
    throw new Error(`Invalid number: ${num}`);
  }
  const numGreen = Math.floor(num / 2);
  const numYellow = num % 2;
  return (
    emojiSquares.green.repeat(numGreen) +
    emojiSquares.yellow.repeat(numYellow) +
    padChar.repeat(5 - (numGreen + numYellow))
  );
}

export function encodeEmoji(emoji: Array<string>, values: Record<string, number>): number {
  const base = Math.max(...Object.values(values)) + 1;
  return emoji.reduce((sum, numEmoji, idx) => sum + values[numEmoji] * base ** idx, 0);
}

export function decodeEmoji(
  encoded: number,
  lookup: Array<string>,
  length: number = -1,
): Array<string> {
  const base = lookup.length;
  const ret: Array<string> = [];
  while (encoded > 0) {
    ret.push(lookup[encoded % base]);
    encoded = Math.floor(encoded / base);
  }
  if (length > 0) {
    if (ret.length > length) {
      throw new Error(`Encoded emoji too long: ${ret}`);
    }
    while (ret.length < length) {
      ret.push(lookup[0]);
    }
  }
  return ret;
}

export function emojiRowsToB64(args: {
  emojiLines: Array<Array<string>>;
  values: Record<string, number>;
  rowsPerB64Word: number;
  b64WordWidth?: number;
}): string {
  const rows = args.emojiLines;
  if (!rows.length) {
    return "";
  }
  const base = Math.max(...Object.values(args.values)) + 1;
  const b64WordWidth = args.b64WordWidth ?? DEFAULT_WORD_WIDTH;
  const rowOffset = base ** rows[0].length;

  if (args.rowsPerB64Word * rowOffset ** args.rowsPerB64Word > 64 ** b64WordWidth) {
    throw new Error(
      `Cannot encode ${args.rowsPerB64Word} rows of ${rows[0].length} emoji (base ${base}) into ${b64WordWidth} b64 chars`,
    );
  }
  let encodedStr = "";
  let currentCode = 0;
  for (let rowIdx = 0; rowIdx < rows.length; ++rowIdx) {
    const positionInRow = rowIdx % args.rowsPerB64Word;
    const isRowEnd = positionInRow === args.rowsPerB64Word - 1;
    const row = rows[rowIdx];
    const rowValue = encodeEmoji(row, args.values);
    currentCode += rowValue * rowOffset ** positionInRow;
    if (isRowEnd || rowIdx === rows.length - 1) {
      // Encode the number of rows in this block.
      currentCode = currentCode * args.rowsPerB64Word + positionInRow;
      encodedStr += intToBase64(currentCode, b64WordWidth);
      currentCode = 0;
    }
  }
  return encodedStr;
}

export function b64ToEmojiRows(args: {
  serializedResult: string;
  lookup: Array<string>;
  rowLength: number;
  rowsPerB64Word: number;
  b64WordWidth?: number;
}): { rows: Array<string>; serializedResult: string } {
  const b64WordWidth = args.b64WordWidth ?? DEFAULT_WORD_WIDTH;
  const base = args.lookup.length;
  const rowOffset = base ** args.rowLength;

  const rows: string[] = [];
  let intResult: number;
  let serializedResult = args.serializedResult;
  while (serializedResult.length > 0) {
    [intResult, serializedResult] = nextInt(serializedResult, b64WordWidth);
    const numRows = (intResult % args.rowsPerB64Word) + 1;
    intResult = Math.floor(intResult / args.rowsPerB64Word);
    for (let rowIdx = 0; rowIdx < numRows; ++rowIdx) {
      rows.push(decodeEmoji(intResult % rowOffset, args.lookup, args.rowLength).join(""));
      intResult = Math.floor(intResult / rowOffset);
    }
  }
  return { rows, serializedResult };
}
