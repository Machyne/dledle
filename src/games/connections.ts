import { BaseGame, GameScore } from "../baseGame";
import { DEFAULT_WORD_WIDTH, intToBase64, nextInt } from "../util/base64";
import { emojiToRegexUnion, listToValueMap, splitEmojiLines } from "../util/emoji";
import { b64ToEmojiRows, emojiRowsToB64 } from "../util/gameHelpers";

const rowsPerB64Word = 2;
const categories = ["ðŸŸ¨", "ðŸŸ©", "ðŸŸ¦", "ðŸŸª"];
const categoryValues = listToValueMap(categories);

export class Connections extends BaseGame {
  get name(): string {
    return "Connections";
  }

  get link(): string {
    return "https://www.nytimes.com/games/connections";
  }

  _buildResultRegex(): RegExp {
    const fourHints = emojiToRegexUnion(categories) + "{4}";
    return new RegExp(`Connections\\s*Puzzle #(\\d+)\\s*((?:${fourHints}\\s*){4,7})`, "u");
  }

  serializeResult(gameResult: RegExpMatchArray) {
    const [, gameNumber, guessesBlock] = gameResult;
    // Next encode the guesses.
    let encoded = "";
    encoded += intToBase64(parseInt(gameNumber), 2);
    const guesses = splitEmojiLines(guessesBlock);
    const winIndexes = categories.map((category) =>
      guesses.findIndex((row) => row.every((color) => color === category)),
    );
    // The guess lines is at most len 7, so elements of winIndexes range from -1 to 6.
    // Increment each index by 1 to shift to range 0 to 7, then encode as a 3-bit number.
    // Four 3-bit numbers can be encoded in two base64 characters.
    const winIndexesEncoding = winIndexes.reduce(
      (sum, winIndex, idx) => sum + (winIndex + 1) * (1 << (3 * idx)),
      0,
    );
    encoded += intToBase64(winIndexesEncoding, 2);

    const wrongGuesses = guesses.filter((_, idx) => !winIndexes.includes(idx));
    // Each square is one of four colors, so we can encode each row as 4 x 2-bit numbers.
    // Two of these 8bit rows + 1 bit row offset (17bits) can be encoded in 3 base64 characters (18bits).
    encoded += emojiRowsToB64({
      emojiLines: wrongGuesses,
      values: categoryValues,
      rowsPerB64Word,
    });
    // If all categories were found, then it's a win. If two were found and one had all but one square, it's a near win.
    const missedCategories = categories.filter((_, idx) => winIndexes[idx] === -1);
    const score =
      missedCategories.length === 0
        ? GameScore.Win
        : missedCategories.length == 2 &&
            wrongGuesses.some((row) =>
              missedCategories.some((miss) => row.filter((color) => color === miss).length === 3),
            )
          ? GameScore.NearWin
          : GameScore.Loss;
    return { score, serializedResult: encoded };
  }

  deserialize(serializedResult: string) {
    if (serializedResult.length < 4 || (serializedResult.length - 4) % DEFAULT_WORD_WIDTH !== 0) {
      throw new Error(`Invalid Connections result: ${serializedResult}`);
    }
    let intResult: number;
    [intResult, serializedResult] = nextInt(serializedResult, 2);
    const gameNumber = intResult;
    [intResult, serializedResult] = nextInt(serializedResult, 2);
    const winIndexes: Array<number> = [];
    for (let i = 0; i < 4; ++i) {
      const idx = (intResult >> (3 * i)) & 0b111;
      winIndexes.push(idx - 1);
    }
    const { rows } = b64ToEmojiRows({
      serializedResult,
      lookup: categories,
      rowLength: 4,
      rowsPerB64Word,
    });
    const winsInOrder = winIndexes
      .reduce(
        (acc, winIdx, idx) => {
          if (winIdx != -1) {
            acc.push([winIdx, categories[idx]]);
          }
          return acc;
        },
        [] as Array<[number, string]>,
      )
      // Sort ascending by win index so that we can insert the earlier rows first and
      // preserve placement for later indexes.
      .sort(([a], [b]) => a - b);
    for (const [winIdx, category] of winsInOrder) {
      rows.splice(winIdx, 0, category.repeat(4));
    }
    return `Connections\nPuzzle #${gameNumber}\n${rows.join("\n")}`;
  }
}
