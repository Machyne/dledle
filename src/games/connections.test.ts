import { GameScore } from "../baseGame";
import { runGameTests } from "../util/testUtils";
import { Connections } from "./connections";

const sampleLateWin = `
Connections
Puzzle #252
游릵游릳游릵游릴
游릳游릳游릳游릳
游릱游릴游릵游릱
游릱游릱游릱游릱
游릴游릵游릵游릴
游릴游릴游릴游릴
游릵游릵游릵游릵
`;

const sampleEarlyWin = `
Connections
Puzzle #252
游릳游릳游릳游릳
游릴游릴游릴游릴
游릱游릱游릱游릱
游릵游릵游릵游릵
`;

const sampleNearWin = `
Connections
Puzzle #252
游릳游릴游릵游릳
游릱游릱游릱游릱
游릳游릳游릴游릳
游릳游릳游릳游릳
游릴游릵游릴游릵
游릴游릴游릵游릴
`;

const sampleFakeNearWin = `
Connections
Puzzle #252
游릳游릴游릵游릳
游릱游릱游릱游릱
游릳游릳游릴游릳
游릳游릳游릳游릳
游릴游릵游릴游릵
游릴游릵游릵游릴
`;

const sampleLoss = `
Connections
Puzzle #252
游릳游릴游릵游릳
游릳游릴游릵游릱
游릴游릳游릵游릱
游릴游릳游릵游릱
`;

describe("Connections", () => {
  runGameTests(new Connections(), [
    {
      name: "sampleLateWin",
      input: sampleLateWin,
      expectedScore: GameScore.Win,
    },
    {
      name: "sampleEarlyWin",
      input: sampleEarlyWin,
      expectedScore: GameScore.Win,
    },
    {
      name: "sampleNearWin",
      input: sampleNearWin,
      expectedScore: GameScore.NearWin,
    },
    {
      name: "sampleFakeNearWin",
      input: sampleFakeNearWin,
      expectedScore: GameScore.Loss,
    },
    {
      name: "sampleLoss",
      input: sampleLoss,
      expectedScore: GameScore.Loss,
    },
  ]);
});
