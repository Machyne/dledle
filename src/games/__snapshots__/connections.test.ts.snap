// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connections serialized data should match snapshot 1`] = `
{
  "sampleEarlyWin": {
    "score": "Win",
    "serializedResult": "8DRj",
  },
  "sampleFakeNearWin": {
    "score": "Loss",
    "serializedResult": "8DECpBC7uP",
  },
  "sampleLateWin": {
    "score": "Win",
    "serializedResult": "8Dy8nzW6DA",
  },
  "sampleLateWin trimmed": {
    "score": "Win",
    "serializedResult": "8Dy8nzW6DA",
  },
  "sampleLateWin with extra link text": {
    "score": "Win",
    "serializedResult": "8Dy8nzW6DA",
  },
  "sampleLoss": {
    "score": "Loss",
    "serializedResult": "8DAAphWjNW",
  },
  "sampleNearWin": {
    "score": "NearWin",
    "serializedResult": "8DECpBC7uO",
  },
}
`;
